src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -0,0 +1,557 @@
+                {
+                    // (2) haven't been visited already
+                    const auto actionID{ actionHasher(actionAndArgs) };
+                    if (visitedActionIDs.find(actionID) == visitedActionIDs.end())
+                    {
+                        const auto& cmdImpl{ get_self<Command>(cmd) };
+                        for (const auto& keys : cmdImpl->KeyMappings())
+                        {
+                            // (3) haven't had that key chord added yet
+                            if (globalHotkeys.find(keys) == globalHotkeys.end())
+                            {
+                                globalHotkeys.emplace(keys, cmd);
+                            }
+                        }
+
+                        // Record that we already handled adding this action to the NameMap.
+                        visitedActionIDs.emplace(actionID);
+                    }
+                }
+            }
+            _GlobalHotkeysCache = single_threaded_map<Control::KeyChord, Model::Command>(std::move(globalHotkeys));
+        }
+        return _GlobalHotkeysCache.GetView();
+    }
+
+    com_ptr<ActionMap> ActionMap::Copy() const
+    {
+        auto actionMap{ make_self<ActionMap>() };
+
+        // copy _KeyMap (KeyChord --> ID)
+        actionMap->_KeyMap = _KeyMap;
+
+        // copy _ActionMap (ID --> Command)
+        for (const auto& [actionID, cmd] : _ActionMap)
+        {
+            actionMap->_ActionMap.emplace(actionID, *(get_self<Command>(cmd)->Copy()));
+        }
+
+        // copy _ConsolidatedActions (ID --> Command)
+        for (const auto& [actionID, cmd] : _ConsolidatedActions)
+        {
+            actionMap->_ConsolidatedActions.emplace(actionID, *(get_self<Command>(cmd)->Copy()));
+        }
+
+        // copy _NestedCommands (Name --> Command)
+        for (const auto& [name, cmd] : _NestedCommands)
+        {
+            actionMap->_NestedCommands.Insert(name, *(get_self<Command>(cmd)->Copy()));
+        }
+
+        // repeat this for each of our parents
+        FAIL_FAST_IF(_parents.size() > 1);
+        for (const auto& parent : _parents)
+        {
+            actionMap->_parents.emplace_back(parent->Copy());
+        }
+
+        return actionMap;
+    }
+
+    // Method Description:
+    // - Adds a command to the ActionMap
+    // Arguments:
+    // - cmd: the command we're adding
+    void ActionMap::AddAction(const Model::Command& cmd)
+    {
+        // _Never_ add null to the ActionMap
+        if (!cmd)
+        {
+            return;
+        }
+
+        // Handle nested commands
+        const auto cmdImpl{ get_self<Command>(cmd) };
+        if (cmdImpl->IsNestedCommand())
+        {
+            // But check if it actually has a name to bind to first
+            const auto name{ cmd.Name() };
+            if (!name.empty())
+            {
+                _NestedCommands.Insert(name, cmd);

```

Comment:
If somebody calls AddAction after either NameMap or NestedCommands or whatever, the cache will be invalid (but you're not invalidating it)